Move Skulpt to Python 3 grammar

meredydd opened the pull request:
Hi folks,

This PR moves Skulpt to the Python 3.7.3 grammar, while maintaining a Python-2 compatibility mode that still passes all our tests. This is the first (and biggest) step towards supporting Python 3 properly in Skulpt.

This is a big change, and I'd like to take a moment to talk about how we're doing it, and why we've chosen to do it this way.

**Background - three awkward truths:**

 1. Skulpt needs to support Python 3, properly.

 2. Many of Skulpt's users need to continue to support Python 2 code (eg Anvil's users have Python 2 code in production; Trinket have a ton of existing Python 2 trinkets out there.)

 3. If we fork the project, the ABIs will diverge immediately, and this will leave Anvil + Trinket (+ probably others) in the dust, stranded on a Python 2 branch


**The philosophy of this upgrade:**
 - So, we're keeping it as one project, in the same way as we've been doing so far.

 - Until now, we've had "Python 3 mode" implemented on top of a Python 2 grammar; from now, we'll have "Python 2 mode" implemented on top of a Python 3 grammar.

 - This is a trade-off: it will make it harder to approach perfect Python 2 support.

 - However, our goal is no longer perfect Python 2 support - it is to *keep everything that currently works working*. This is much more achievable.

 - It's easier (and saner) than you might think. The differences are mostly pretty small - retrofitting Python 2 syntax to the grammar was pretty straightforward. The original motivation behind Python's 2.7->3.0 transition was making strings Unicode, and this is:
   * (a) something where we currently have sufficiently poor fidelity to Python 2 that there's not a lot of API surface area to be backward-compatible with, and
   * (b) mostly a stdlib change, with a tiny bit of cooperation from the compiler, and keeping a parallel string class around is *vastly* easier than keeping a parallel parser/AST/compiler around


**What we've done:**
 - We have moved the parser and compiler over to using the Python 3.7.3 grammar and AST. (This is the "big bang" of the Python 3 upgrade, because we can't sensibly have two compilers in the same Skulpt version)

 - We have retrofitted support for Python-2-specific syntax (eg the `print` statement)

 - We have got all the existing run-tests and unit tests working

 - We have implemented a couple of Python-3-specific features (rvalue tuple expansion, eg `x = (a,**b,c)`, and keyword-only arguments, eg `def myfun(a, *, b):`), to demonstrate that it's possible, and how version-specific syntax is controlled by the `__future__` flags.


**What we _haven't_ done:**
 - Updated any of the stdlib (eg `dict.items()` still returns a list in both Python 2 and Python 3 mode)

 - Implemented most Python 3 features and syntax (eg no f-strings quite yet)

 - Implemented `unicode`

 - Made Closure happy yet (coming very soon!)


**What we'd recommend:**
 - I'd suggest doing a stable release of Skulpt before this change gets merged.

 - We'll want to get everyone testing this and reporting/fixing anything that's been broken in their Python 2 code.

 - Now we've opened the floodgates, we can start working on Python 3 features! But please, let's avoid breaking the Python 2 users in the process - and keep this in mind when reviewing PRs, etc.

----

I'm really excited about this. I think this is the "big hump" - with this change merged, we can bring all our Python 3 features online, piece by piece, while keeping Skulpt usable for all our existing projects.

@skulpt/committers, comments please!

Meredydd & Albert-Jan

rixner commented:
Very exciting!

I haven't looked at the code, but I just ran the tests for sanity, and I get this:

```
test/unit/test_time.py

Fail: Expected Jan 01 1970 07:01:01 to equal Jan 01 1970 01:01:01
Tests failed in  TimeTestCase.test_strftime  
Ran 6 tests, passed: 5 failed: 1
```

albertjan commented:
I'm so happy! I started this work almost two years ago! But I knew then that I'd need someone to help me pull this all the way through the end. @meredydd has taken this from maybe a way forwards to actually a viable plan. 

There are still a few outstanding things, but we felt that with all the tests passing it was a good time to open it up for comments. 😄

meredydd commented:
@rixner - Oh, no! I thought I'd fixed that for everyone, not just me ;)

In any case, this is an issue with time.js and `test_time`, and not related to our changes. The version on master is broken for @albertjan and me (in the UK, which is currently on GMT+1 summer time), which is why I wrote commit 4ace104, but clearly that's broken it for you.

rixner commented:
@meredydd I know.  That test_time file is a pain because we are all in different time zones, think we fixed it globally, and then someone else finds out we have not.  You did change the test code, so in that sense it is "related" to your changes. :)  But, it clearly has no bearing on the actual meat of this pull request.  We just need to figure out how to fix it correctly so that we can all pass that test.

I have read through a lot of the code and am happy so far.  I want to run it and read it some more, though, as it is a lot to digest.  But, I definitely do not want this languishing here for weeks as we debate it.  I agree wholeheartedly with your plan that we make another stable release before accepting this, but then once we've all digested it a little, we accept it and deal with the consequences as we fight any fires that come up.

And to be clear, from my end, I don't care **at all** about Python 2 anymore.  I will be supporting Python 2 indefinitely, but I am willing to freeze my Python 2 support at a Skulpt version that has already been released.  Which is not to say that I will break your Python 2 support, I will not (at least not intentionally!).  Just putting that out there as a counter-point.

This is a **huge** step forward.  Nicely done guys!

bnmnetp commented:
Wow!  I'm in awe of you guys.  @albertjan @meredydd 

Thanks for all the hard work!

P.S. Take your time on the dict.items() issue.  I would be fine if it never changed to some kind of mysterious dict_view.  Its so much easier to explain a list of tuples to novice programmers 😄

rixner commented on src/compile.js:
```diff
@@ -878,6 +951,11 @@ Compiler.prototype.vseqexpr = function (exprs, data) {
     Sk.asserts.assert(data === undefined || exprs.length === data.length);
     ret = [];
+
+    // if (exprs.length === 1 && exprs[0].constructor === Sk.astnodes.Starred) {
```
It seems like these sorts of commented blocks end up staying in the code forever.  Can we either delete such things before merging or put a descriptive comment explaining what this is so that someone in the future can fix and uncomment it (if that's what needs to be done)?

rixner commented:
I obviously want to see what happens with the closure compiler, but one thing that makes me sad is that I'm seeing a 1.5-2x slowdown with this. :frowning_face:

rixner commented:
I have read a large fraction of the diff.  Obviously the broken time test is going to need to be fixed somehow and the closure compiler made happy.  I'd like to better understand where the slowdown might be coming from (and if it still exists after closure compilation - I'm comparing devbuild in master to devbuild on this branch running the tests, which is not a comprehensive performance evaluation by any means, but I don't want to spend too much time on that until I see the optimized build).

Having said that, other than the few really minor comments I've left, I don't find anything else that looks concerning to me!  I want to run more of my code through this still, but wanted to leave a positive statement about this for now (not just leave it with my performance woes)! :smile:

albertjan commented:
> I obviously want to see what happens with the closure compiler, but one thing that makes me sad is that I'm seeing a 1.5-2x slowdown with this. 

Performance hasn't been on our minds while developing this. :smile: I'm sure we can find a few bottle necks to close the gap.

meredydd commented:
Re performance, my first idea would be to look at the code generation for `Call` nodes, which has slowed down a fair bit. Even before we started trying to handle things like multiple `**kwdict` args, we had a few correctness issues, so I went for "dumb but correct" until I could actually get it working. The code we generate there is certainly not optimal. (We also now allow for suspensions when traversing `**kwdict` arguments, which probably costs us a fair bit.)

Likewise, there's been a full rewrite of `Sk.builtin.func.tp$call()`, triggered by the addition of Python 3's keyword-only arguments. It's possible that something I've done there has tickled the optimiser wrong.

Overall, because the actual code is so similar (and the tests haven't really change), I think we could get good mileage out of running unit tests on a profiler right before and right after this merge -- they'll be pretty directly comparable, so we should be able to see where the slowdown came from.

But, given the magnitude of the changes here, I would suggest merging this (and putting out what fires it inevitably lights) first!

albertjan commented on src/env.js:
```diff
@@ -33,6 +33,7 @@ Sk.python2 = {
     absolute_import: null,
     unicode_literals: false,
     // skulpt specific
+    python3: false,
```
so there are a few things that need their own flag because they are importable from `__future__` and when we only had a few python3 features it made sense from a production environment to be able to turn them on only for new stuff so you don't break old code out there. But since there are now probably more python2 versions you want to turn off instead of python3 version to turn on. We felt this would do for now and we see if it makes sense to switch every difference between py3 and py2.

rixner replied:
So moving forward, we can/should just use the python3 future flag for things we add instead of creating new future flags?

meredydd replied:
Yes, unless there's a particular reason to make it switchable independently.

rixner replied:
Sounds good!

albertjan commented:
re: tp$call performance: cpython has a fast call route that skips all the argument mangling if the number arguments specified equals parameters of the function.

rixner commented:
The strftime time test still needs to be fixed.  I believe the original assert is correct, the new assert is incorrect, and the bug is in either localtime and/or timezone.

We could decouple this issue from this pull request, though, by commenting out the strftime test...

meredydd commented:
_Re the time tests:_
The original assert is incorrect (`time.timezone` is specified as "the offset of the local (non-DST) timezone", so that assert _will_ fail during DST). I've changed the test to check whether we're currently in DST, and switch between `time.timezone` and `time.altzone`; hopefully this is the One True Correct solution (within the limited concept of timezones and DST supported by the `time` module).

_Re tp$call performance:_
First, I strongly believe we shouldn't hold up the merge on this -- there's a lot of (parallelisable) work for lots of people that this will unlock, and performance.

Second, a word of caution -- that code has been broken (several times?) before, presumably in the pursuit of speed -- which is one reason why I reverted to a pretty faithful port of `_PyEval_EvalCodeWithName`.

That said, I do like @albertjan's suggestion of simple fast paths for the common case. That seems both maintainable and an improvement.

meredydd commented on src/compile.js:
```diff
@@ -2330,15 +2444,6 @@ Compiler.prototype.nameop = function (name, ctx, dataToStore) {
         throw new Sk.builtin.SyntaxError("can not assign to None");
     }
 
-    if (name.v === "None") {
```
Replaced with an assertion, and better / more explicit errors introduced.

rixner replied:
I'm not seeing that change.  Did it not get pushed?

meredydd replied:
https://github.com/skulpt/skulpt/pull/905/commits/d1cfb7d4f7c8f1f984489b4a96660025d5cd6bc4

rixner replied:
Got it, thanks.

rixner commented:
Running in Chrome and looking at the profile, it appears that literally everything is just slower.  All the functions in the profile show up as taking longer; compiled, internal, etc.

Looking at the v8 profile locally, the issue appears to be the "IC" functions.  Which I believe are the inline cache?  For example, on one profiling run, "LoadIC" took up 22.4% of the time in this version (compared to 2% in master) on a program I ran...  In this program, this version takes 1.6 times as long as master, so LoadIC by itself is a lot of it, but not all.  Interestingly, when I run on Chrome, it takes only 1.2 times as long, which is much more palatable.

It seems like we would need to understand why this is interacting poorly with the v8 cache in order to knock back the performance loss.  (Or I could be misreading/misunderstanding the profile.)  I also haven't tried in Firefox or Safari...

Browser: Chrome 75.0.3770.90 (V8 7.5.288.23)
Local: node v8.10.0

LeszekSwirski commented:
`LoadIC` is specifically performing the load operation for named properties (`x.foo`). This typically gets slower if you do things like passing differently shaped objects into the load, where it can't cache the shape of the object & the expected offset of the field on the object. Another possibility is that objects are transitioning from "fast" to "slow/dictionary" mode, e.g. through use of `delete`. Or, it could just be taking up more of the profile because there are more loads happening...

albertjan commented:
@rixner I'm hope any performance loss is made up by the step forward this PR does. We have already identified at least one avenue to address the performance. It feels like we're focusing on the wrong thing here a little bit 😄. 

We can address the performance issues in separate PR's. This one is already quite big.

rixner commented:
@albertjan I'll say 2 things:

1)  I respectfully disagree that I'm focusing on the wrong thing.   I believe this is great work and said so in my initial comments.  I still believe that and hope I did not give the impression I think otherwise!  However, history has established that I care far more about performance than anyone else. And once performance-hurting pull requests are accepted, others appetites for worrying about it goes down.  So I think now is exactly the time to talk about it. 

2) I am now of the opinion that the benefits outweigh the costs at around 20% slowdown.  2x was just disheartening to me, but my conclusion is much of that is compilation time on short programs. And the 1.5x seems to be on older v8 versions.

I would like to run some more programs through to confirm functionality, but then intend to submit an approving review. Hopefully sometime tonight.

rixner commented:
```
test/unit/test_time.py

Fail: Expected Jan 01 1970 00:01:01 to equal Jan 01 1970 01:01:01
Tests failed in  TimeTestCase.test_strftime  
Ran 6 tests, passed: 5 failed: 1
```

I don't think you can check if isdst is true right now and then apply that to some time in the past.  You need to check it for the time in the past that you want to use it with.

CPython 3:

```
Python 3.6.8 (default, Jan 14 2019, 11:02:34) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> t = time.localtime(3661)
>>> t
time.struct_time(tm_year=1969, tm_mon=12, tm_mday=31, tm_hour=19, tm_min=1, tm_sec=1, tm_wday=2, tm_yday=365, tm_isdst=0)
>>> t2 = time.localtime()
>>> t2
time.struct_time(tm_year=2019, tm_mon=6, tm_mday=17, tm_hour=22, tm_min=25, tm_sec=40, tm_wday=0, tm_yday=168, tm_isdst=1)
>>> time.timezone
21600
>>> time.altzone
18000
>>> is_dst = time.localtime().tm_isdst == 1
>>> is_dst
True
>>> tz = time.altzone if is_dst else time.timezone
>>> tz
18000
>>> time.strftime("%b %d %Y %H:%M:%S", time.localtime(3661 + tz))
'Jan 01 1970 00:01:01'
>>> 
```

albertjan commented:
> I respectfully disagree that I'm focusing on the wrong thing. I believe this is great work and said so in my initial comments. I still believe that and hope I did not give the impression I think otherwise! However, history has established that I care far more about performance than anyone else. And once performance-hurting pull requests are accepted, others appetites for worrying about it goes down. So I think now is exactly the time to talk about it.

You're absolutely right to focus on the performance and we have forgotten about it on many occasions 😄 I myself have definitely been the cause of some of those slowdowns. I've also found it's really difficult to properly profile javascript (in browser of on node). I was just afraid to add to this already massive PR if we were to hold off merging this till those performance issues are fixed.

I'm willing to focus some time on the performance of skulpt for a while there must be some wins to be had. Like the early out of `tp$call`. It might be good to have a look at cpython's default `tp$getattr` implementation, they must do some optimisation there too. We probably also introduced a slowdown when I implemented function descriptors. Each function call on an class now has an extra call added to it to bind it to the class. It's of course what python is supposed to do but it probably does slow things down. But with these things it's again hard to see how we could optimise it. 

I guess my comment earlier was more to make sure we don't forget to focus on all of the other changes too. 

Performance is such a tricky moving target, if you point at a line of code and say this is wrong, or you could have done this differently. It's much easier to address. 😄 

@rixner Do you still use `field.py` to do the profiling?

We should all just move to UTC land. 🤦‍♂ And abolish dst.

bnmnetp commented:
Hi All,

So, submitted (jointly) by @meredydd and @albertjan, approved by @rixner 

I did a tagged release not tool long ago, just before the move to node.  Other than a few bug fixes on turtle, I don't think we've had anything significant other than our own infrastructure updates.

I'm happy to pull the trigger on this to get it merged so we can move on to performance and other feature updates.  But there was some discussion about making another release.  That seems unnecessary to me given that the changes were almost entirely around the move to node.  But its easy enough to make a tag....

Comments?

meredydd commented:
The time thing gets weirder and weirder; `time.localtime()` is not doing what either of us think it is. Let's separate that conversation: I have opened #908 to track it.

meredydd commented:
> But there was some discussion about making another release. That seems unnecessary to me given that the changes were almost entirely around the move to node. But its easy enough to make a tag....

My vote: I think the recent tooling changes on their own deserve at least a minor release. (Plus, looking through the commit history, I see a steady flow of other fixes!)

bnmnetp commented:
Fair enough, I'll go ahead and tag what we have 1.1 -- The changes to tooling probably do deserve a bump before we move on.

There are two smallish PR's outstanding 

The changes too processing.js which I think I can just accept, and turtle.js. @albertjan are you good to go with that one?

rixner commented:
@albertjan I still do use field.py, but that doesn't exhibit the bad behavior here.  Unfortunately, I can't share that code, as it's part of my classes and I don't want it public.  If I can figure out the underlying structures that evoke the problem, hopefully I can find or build a useful benchmark program to share.  And I agree that we should abolish DST, but I'm not so sure about moving to UTC land. ;)

@bnmnetp I am fine with or without a release.  However, I don't think this should be merged until the time test is fixed, commented out, or whatever.  #908 suggests there might not be a quick fix.

bnmnetp commented:
@rixner -- I thought the time test had been commented out.  I don't think we should hold up this PR over a problem that has existed for ages. It bit me last spring in Dominican Republic where they do not observe DST.   I agree with @meredydd that we should separate it and track it separately.

rixner commented:
@bnmnetp it wasn't.  @meredydd just "fixed" it by changing the test to something that should always pass for all of us.  (Hopefully!)

meredydd commented:
If you like watching programmers give themselves migraines, a read of the #908 thread may entertain you. tl;dr timezones are complicated.

As for this one, though, let's merge it!

rixner commented:
I'm with @bnmnetp , I think #906 and #901 should be merged first, then the version bumped, then this one.  Presumably the other two can be dealt with quickly.

bnmnetp commented:
So, I have tagged the state of things at 1.1.0 (not including this PR)

## Differences from 1.0.0

* New tool set that uses node and webpack
* closure compiler outputs ES-2015
* Fixes to turtle and processing
* General cleanup / bugfixes

----

Read the description above to understand the big leap this PR makes.

Thanks again to all!

acbart commented:
Just to tell you all things you already know, it does appear that the new Skulpt version is much slower. BlockPy has a library for analyzing student code (Pedal), which is a fairly hefty thing. Previously, it took about a full second or two to run in Skulpt. Right now it's taking about 20 seconds to simply import it (10 seconds if I precompile it). Hopefully there'll be some time and attention paid to performance at some point. Still, great progress with the new features!
